docs/specifications/tests/Syntax-TP/syntax06.md

```diff
@@ -1,54 +1,129 @@
+in section 3.3.13 of [RFC 1035] and section 2.2 of [RFC 1912]. The RNAME
+field should follow the rules of an e-mail address also defined in 
+[RFC 5322, section 3.4.1].
+
+## Inputs
+
+* "Child Zone" - The domain name to be tested.
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Obtain the set of name server IP addresses using [Method4] and [Method5]
+   ("Name Server IP").
+
+2. Create a SOA query for the apex of the *Child Zone* with RD flag unset.
+
+3. For each name server in *Name Server IP* do:
+   1. Send the SOA query over UDP to the name server.
+   2. If the name server does not respond with a DNS response, then:
+      1. Output *[NO_RESPONSE]*.
+      2. Go to next server.
+   3. If the DNS response does not include a SOA record in the 
+      answer section, then:
+      1. Output *[NO_RESPONSE_SOA_QUERY]*.
+      2. Go to next server.
+   4. Extract the RNAME from the SOA record.
+   5. Convert the first "." without backslash quoting to an "@" in 
+      the RNAME.
+   6. Convert any backslash quoted "." to a single "." without quoting
+      (see section 5.1, 5.3 and 8 in [RFC1035] for the use of backslash).
+   7. If the converted string (mail address) does not meet the 
+      specification in [RFC 5322, section 3.4.1], then 
+      1. Output *[RNAME_RFC822_INVALID]*.
+      2. Go to next server.
+   8. Extract the domain part (to the right of "@") from the mail 
+      address ("Domain Part").
+   9. Create an MX query for the *Domain Part* and send it to a resolving
+       name server. 
+   10. If the MX lookup does not return a DNS response with RCODE 
+       "NOERROR", then:
+       1. Output *[RNAME_MAIL_DOMAIN_INVALID]*.
+       2. Go to next server.
+   11. If the MX lookup returned a CNAME or a chain of CNAMEs then
+       set *Domain Part* to be the the final name instead of the
+       domain from the email address.
+   12. If the MX lookup returned a NO DATA response (no MX record), 
+       then:
+       1. Create address queries (A and AAAA) for the *Domain Part*
+          and send it to a resolving name server.
+       2. Disregard any A record with 127.0.0.1 or AAAA with ::1.
+       3. If no A or AAAA records with the same owner name as *Domain
+          Part* were found in the responses 
+          then output *[RNAME_MAIL_DOMAIN_INVALID]*.
+   13. If the MX lookup returned one or more MX records, then for each
+       mail exchange (domain name in RDATA of the MX record) do:
+       1. Create address queries (A and AAAA) and send it to a 
+          resolving name server.
+       2. Disregard all A and AAAA records that do not have the same
+          owner name as the mail exchange name (i.e. do not follow
+          any CNAME).
+       3. Disregard any A record with 127.0.0.1 or AAAA with ::1.
+       4. If all MX have been processed and neither A or AAAA record 
+          was returned for any mail exchange, then output 
+          *[RNAME_MAIL_DOMAIN_INVALID]*.
```

Comment:
I really think it's improper to check if the foreach-loop has completed its work from within the foreach loop itself.
The problem could be solved using much simpler constructs. E.g. by introducing a single boolean variable. The variable could be initialized to false before the loop, set to true when relevant addresses are found inside the loop, and checked after the loop to determine whether RNAME_MAIL_DOMAIN_INVALID is to be emitted.

Reply:
For a computer program I agree, but for a human program I think the current solution is simpler.

Reply:
I disagree with the notion of "human programs". All the programs we care about are maintained by humans.

Reply:
Found a solution without control variable. The text is written to be read, not to be executed.